### ISSUE TITLE
Adapting hist_local to 16bit images - cannot understand indexing

### ISSUE THREAD

**Issue by indridieinarsson** [libvips/libvips#1460]
I'm trying to adapt the CLAHE method in hist_local.c to also allow 16 bit images (I have a bunch of scanned b&w film in 16bits that I'm playing with, don't want to convert to 8bits).

I've only done the obvious things : replace the hard-coded occurences of 255 and 256 with 65535 and 65536, resp. in order to use 16bit histograms instead of 8bits. Then to change the check for `UCHAR` to also allow `USHORT.` 
So far so good. When rolling this through nip2 on a `GREY16` image, the result looks like it is the correctly filtered image, but with vertical black stripes on it. I get one 32pixels wide stripe of image, then 32px wide black stripe, then 32bits image etc.

I haven't been able to find other occurrences of uchar being implied in this code (and nip2 doesn't seem to do so either), but the 32pixel alternating bands indicate that this is probably the case.

I'm not very "fluent" in reading c code (bunch of fortran experience though ;) ), and realize the issue is surely on my side of the keyboard. Could someone point me in the right direction, e.g. where in this file 8bit images are being (perhaps silently) assumed, or point to a similar code that I can mimick.

Greetings,
Indriði

**Comment by indridieinarsson** [comment]
I've been trying to understand the innards of libvips, have made some progress understanding, but still get weird stuff when running hist_local on 16bit grayscale images. I just swapped 255 for 65535, 256 for 65536, UCHAR for USHORT in the check. I should not have to change other types, I think, since unsigned int is being used all over, and that has enough bits to store a sum over unsigned short (16 vs 32 bits). Nowhere is unsigned char being used explicitly.

(I will need to implement a binned histogram, as 65536 bins for e.g. 100x100 region make no sense)

The image comes out something like this: 
![normal](https://user-images.githubusercontent.com/4919928/68347355-8dd53c80-00ee-11ea-932e-afd292a25de3.jpg)

I tried changing the line `"extend", VIPS_EXTEND_MIRROR` into `"extend", VIPS_EXTEND_WHITE`, now the result looks like this:
![extend_white](https://user-images.githubusercontent.com/4919928/68347436-d7be2280-00ee-11ea-8e0e-53000e81bbb7.jpg)
(note a 64px wide white stripe on the left hand side of the image).
This must be an indexing issue into the image. Have the feeling that a "strategically located" *2 or /2 will do it, but haven't been able to understand where that should happen. 
Does anyone have a hint for me?

**Comment by jcupitt** [comment]
Hello @indridieinarsson,

You'll need to add a new path to the _generate function for ushort pixels. Have a look at `vips_hist_cum_process` in `hist_cum.c`, for example. 

tldr: libvips needs a separate bit of code in every operation for every numeric format. This is usually generated with macros: you parameterize on the type and expand N times inside a switch.
